# Lab 04: Duplicates

- **Name**:  Colton Crum
- **Netid**: ccrum

## Activity 1: Brute-Force

### Function

In [261]:
def has_duplicates_bf(phrase):
    phrase = phrase.split()
    
    for duplic in phrase:
        word = duplic
        for match in phrase[phrase.index(word) + 1:]:
            if match == word:
                return True
           
    return False

### Testing

In [262]:
LYRICS = [
    "But I've got a blank space, baby and I'll write your name",
    "Cause we never go out of style, We never go out of style.",
    "Baby, I'm just gonna shake, shake, shake, shake, shake, I shake it off, I shake it off",
    "So take a look what you've done, Cause, baby, now we got bad blood",
    "Ooh, look what you made me do, Look what you made me do",
]

In [263]:
def check_lyrics(function, display=True):
    for lyric in LYRICS:
        if display == True:
            print("{} -> {}".format(lyric,function(lyric)))
        else:
            function(lyric)
    
    pass

In [264]:
check_lyrics(has_duplicates_bf, display=True)

But I've got a blank space, baby and I'll write your name -> False
Cause we never go out of style, We never go out of style. -> True
Baby, I'm just gonna shake, shake, shake, shake, shake, I shake it off, I shake it off -> True
So take a look what you've done, Cause, baby, now we got bad blood -> False
Ooh, look what you made me do, Look what you made me do -> True


### Benchmarking

In [265]:
%timeit check_lyrics(has_duplicates_bf, display=False)

38.9 µs ± 661 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


### Reflection

1. Describe the flow control of your `has_duplicates_bf` function.  How did it detect if there was a duplicate word?

    <font color="red">My program split the list into an string, separating each word in the index. This allowed the program to then iterate over the string using a nestled for loop. The for loop returned true when the variable duplic was equal to the variable lyric, which was each word in the string. The second for loop determines if the variable match in the string was equal by comparison to word, then the program found a duplicate. If this was not the case, the program returned false, meaning there was no duplicates.</font>

2. What is the algorithmic complexity of your `has_duplicates_bf` function?

    <font color="red">The complexity of the program is 0(n), since the program is reiterating through the list, checking every element starting from the begining of the string.</font>

## Activity 2: Sort and Scan

### Function

In [266]:
def has_duplicates_sc(phrase):
    phrase = phrase.lower()
    phrase = phrase.split()
    phrase = sorted(phrase)
    
    for word in phrase:
        if (phrase.index(word) < len(phrase) -1) and (word == phrase[phrase.index(word) + 1]):
            return True
        
    return False
    
        

### Testing

In [267]:
check_lyrics(has_duplicates_sc, display=True)

But I've got a blank space, baby and I'll write your name -> False
Cause we never go out of style, We never go out of style. -> True
Baby, I'm just gonna shake, shake, shake, shake, shake, I shake it off, I shake it off -> True
So take a look what you've done, Cause, baby, now we got bad blood -> False
Ooh, look what you made me do, Look what you made me do -> True


### Benchmarking

In [268]:
%timeit check_lyrics(has_duplicates_sc, display=False)

37.4 µs ± 971 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


## Reflection

1. Describe the flow control of your `has_duplicates_sc` function.  How did it detect if there was a duplicate word?
    
    <font color="red">The has_duplicates_sc function first changed the list into a string and all lowercase with phrase.lower(). The string is then split into each word by phrase.split(), and finally sorted with sorted(phrase). Then a for loop checks word in phrase, to see if word is equal to an index in the string. If it is, the function will return True, if not, False.</font>

2. What is the algorithmic complexity of your `has_duplicates_sc` function?

    <font color="red">The algortihmic complexity is 0(n), since the program is run through each index of the list. By sorting each word better than activity one with .lower() and sorted functions, the computer was able to process each word faster, even though both activity 1 and 2 were still algromically the same linear search. Sorting does make a difference in relation to speed for this type of elements or data.</font>

## Activity 3: Sort and Search

### Function

In [269]:
import bisect

def has_duplicates_bs(phrase):
    phrase = phrase.lower()
    phrase = sorted(phrase.split())
    for wordindex in range(len(phrase)):
        found = bisect.bisect_right(phrase, phrase[wordindex])
        if wordindex + 1 != found:
            return True
       
    return False


### Testing

In [270]:
check_lyrics(has_duplicates_bs, display=True)

But I've got a blank space, baby and I'll write your name -> False
Cause we never go out of style, We never go out of style. -> True
Baby, I'm just gonna shake, shake, shake, shake, shake, I shake it off, I shake it off -> True
So take a look what you've done, Cause, baby, now we got bad blood -> False
Ooh, look what you made me do, Look what you made me do -> True


### Benchmarking

In [271]:
%timeit check_lyrics(has_duplicates_bs, display=False)

31.7 µs ± 149 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


## Reflection

1. Describe the flow control of your `has_duplicates_bs` function.  How did it detect if there was a duplicate word?

    <font color="red">The has_duplicates_bs function changes the list into a string, makes every element lowercase with phrase.lower, and split into each index with sorted(phrase.split()). Then the for loop uses forindex in the range of the string, if wordindex is equal to its binary then it will return true, stating a match. If the binary is not the same as the string is iterated, then there is no match for wordindex and found, which holds the counts for what each element in the data set is.</font>

2. What is the algorithmic complexity of your `has_duplicates_bs` function?

    <font color="red">The algorithmic complexity of the has_duplicates_bs is 0(log(n), since this is the average case for binary search. This technique doesn't help our speed for this data set, since a linear search is more straight forward using small data sets.</font>

## Activity 4: Reflection

Examining the results of Activities 1, 2, and 3, what can you say about the relationship between algorithmic complexity and real world performance?

<font color="red">Though these programs function similarly, they complete the task differently and at different speeds. The way these algrioms are structured affect how long they take to complete. With the first activity, the program is moving through each index until it finds a match with a word that it has stored. This was elementary and took the longest to complete, as it was the slowest complexity 0(n) at 38.9 nanoseconds. With activity 2, the program worked similarly to the first, but the index was sorted with lowercase letters and stored the checking varaible differently from its list length. This speed up the algorithim to 37.4 nanoseconds. For activity 3, the index was splited, sorted, and then converted into binary. This allowed wordindex and found to search through the string. This was a lot faster, at 31.7 nanoseconds. Algorithmic complexity changes the speed needed to complete the task, and efficiency. With a different type of data, linear search may be more useful, and at smaller volumes. Each technique has its own place to find and sort through matches of data. </font>